lib/api.js

```diff
@@ -530,6 +536,93 @@ api.declare({
     return x;
   })});
 });
+
+api.declare({
+  method: 'delete',
+  route: '/credentials',
+  name: 'claimCredentials',
+  title: 'Request worker credentials',
+  stability: API.stability.experimental,
+  input: 'claim-credentials-request.json#',
+  description: [
+    'Yield credentials for the worker running in the given instance.',
+  ].join(' '),
+}, async function(req, res) {
+  const reportError = (obj, msg) => {
+    log.error(obj, msg);
+    return res.reportError('AuthorizationFailed', 'Failure to verify authorization');
+  };
+
+  const doc = getIdentityDocument(req.body.signature);
+
+  if (!doc) {
+    return reportError({
+      signature: req.body.signature,
+    }, 'Invalid signature');
+  }
+
+  this.state.claimCredentials({
+    region: doc.region,
+    id: doc.instanceId,
+  });
+});
+
+api.declare({
+  method: 'get',
+  route: '/credentials',
+  name: 'getCredentials',
+  title: 'Request worker credentials',
+  stability: API.stability.experimental,
+  input: 'claim-credentials-request.json#',
+  output: 'claim-credentials-response.json#',
+  description: [
+    'Yield credentials for the worker running in the given instance.',
+  ].join(' '),
+}, async function(req, res) {
+  const reportError = (obj, msg) => {
+    log.error(obj, msg);
+    return res.reportError('AuthorizationFailed', 'Failure to verify authorization');
+  };
+
+  const doc = getIdentityDocument(req.body.signature);
+
+  if (!doc) {
+    return reportError({
+      signature: req.body.signature,
+    }, 'Invalid signature');
+  }
+
+  const instances = await this.state.listInstances({
```

Comment:
Let's move the state-based logic for answering the question of "is this instance in a state to generate credentials" into a `State.canClaimCredentials({id, region}, client)` method.